File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValid(root, float('-inf'), float('inf'))
        
    def isValid(self, node: Optional[TreeNode], lower_bound: float, upper_bound: float) -> bool:
        if not node:
            return True
        if not lower_bound < node.val < upper_bound:
            print(node.val, lower_bound, upper_bound)
            return False
        return self.isValid(node.left, lower_bound, node.val) and self.isValid(node.right, node.val, upper_bound)
        

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int KthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new();
        var node = root;
        while (node != null || stack.Count > 0)   
        {
            while (node != null)
            {
                stack.Push(node);
                node = node.left;
            }

            node = stack.Pop();
            k -= 1;

            if (k == 0)
            {
                return node.val;
            }

            node = node.right;
        }
        return -1;
    }
}
